internal2.lua

```diff
@@ -332,6 +336,10 @@ function Tube:add_tube_dir(pos, dir)
 	if param2 then
 		local d1, d2, num = self:decode_param2(npos, param2)
 		if not num then return end
+		-- if invalid face, do nothing
```

Comment:
the function Tube:add_tube_dir handles tube node only. There is no need for any invalid face check. This is already implemented in param2 (dir1, dir2, num_connections).

Reply:
This piece of code is required to stop tubes that are already placed, but have an end that is not connected, from connecting that end to an invalid side of a face.
Consider:
I place a tube facing north-south
I place a pusher next to it, facing north-south
The tube decides to connect one of its ends to the side of the pusher, even though that cannot be a valid configuration

This code prevents that from happening (so long as valid_sides are properly defined for pushers, as in https://github.com/joe7575/techpack/pull/76/files)

Reply:
Ok, I understand. Good point.

Reply:
If there's a better place to put it let me know, I really am guessing with the positioning and trying not to dig down too far. I'm certain all the functionality is required, but how it's implemented it totally up for debate.